### PR TITLE
add a Dynamic extractor interface

### PR DESCRIPTION
**Pull request by yelhamer**
This is the parent branch of the cape-dynamic-extractor and cuckoo-dynamic extractor; The two will ensue later on. It defined the interface that all dynamic sandbox-based feature extractors should implement.

### Checklist

- [x] No CHANGELOG update needed

- [x] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/features/extractors/base_extractor.py [hunk 1]
```diff
@@ -262,3 +262,105 @@ def extract_insn_features(
...
+        """
+        Yields all the features of a process. These include:
+        - file features of the process' image
+        - inter-process injection
+        - detected dynamic DLL loading
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_threads(self, ph: ProcessHandle) -> Iterator[ProcessHandle]:
```
Comment (mr-tz) [inline capa/features/extractors/base_extractor.py:330]:
```suggestion
    def get_threads(self, ph: ProcessHandle) -> Iterator[ThreadHandle]:
```

File: capa/features/extractors/base_extractor.py [hunk 2]
```diff
@@ -262,3 +262,105 @@ def extract_insn_features(
...
+            ph: parent process
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def extract_thread_features(self, ph: ProcessHandle, th: ThreadHandle) -> Iterator[Tuple[Feature, Address]]:
+        """
+        Yields all the features of a thread. These include:
+        - sequenced api traces
+        - files/registris interacted with
```
Comment (mr-tz) [inline capa/features/extractors/base_extractor.py:344]:
```suggestion
        - file/registry interactions
```

**Review by mr-tz — COMMENTED** [review]
looks promising :) sorry if I jumped into review too quickly, I'm just very excited

File: capa/features/extractors/base_extractor.py [hunk 3]
```diff
@@ -262,3 +262,105 @@ def extract_insn_features(
...
+class ProcessHandle:
+    """
+    reference to a process extracted by the sandbox.
+
+    Attributes:
+        pid: process id
+        inner: sandbox-specific data
+    """
+
+    pid: int
```
[the remark this answers is not in the document]
Comment (williballenthin) [reply capa/features/extractors/base_extractor.py:279]: tracked in #1547

File: capa/features/extractors/base_extractor.py [hunk 4]
```diff
@@ -262,3 +262,105 @@ def extract_insn_features(
...
+    Features are grouped mainly into threads that alongside their meta-features are also grouped into 
+    processes (that also have their own features). Other scopes (such as function and file) may also apply 
+    for a specific sandbox.
+
+    This class is not instantiated directly; it is the base class for other implementations.
+    """
+
+    def __init__(self):
+        super().__init__()
+
```
[the remark this answers is not in the document]
Comment (williballenthin) [reply capa/features/extractors/base_extractor.py:310]: tracked in #1548

File: README.md [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 [![PyPI - Python Version](https://img.shields.io/pypi/pyversions/flare-capa)](https://pypi.org/project/flare-capa)
 [![Last release](https://img.shields.io/github/v/release/mandiant/capa)](https://github.com/mandiant/capa/releases)
-[![Number of rules](https://img.shields.io/badge/rules-799-blue.svg)](https://github.com/mandiant/capa-rules)
+[![Number of rules](https://img.shields.io/badge/rules-800-blue.svg)](https://github.com/mandiant/capa-rules)
 [![CI status](https://github.com/mandiant/capa/workflows/CI/badge.svg)](https://github.com/mandiant/capa/actions?query=workflow%3ACI+event%3Apush+branch%3Amaster)
 [![Downloads](https://img.shields.io/github/downloads/mandiant/capa/total)](https://github.com/mandiant/capa/releases)
 [![License](https://img.shields.io/badge/license-Apache--2.0-green.svg)](LICENSE.txt)
```
Comment (williballenthin) [inline README.md:8]: these extra changes stem from us not keeping the feature branch in sync with master. i'll do a merge from there.

File: capa/features/extractors/base_extractor.py [hunk 6]
```diff
@@ -262,3 +262,88 @@ def extract_insn_features(
...
+    reference to a process extracted by the sandbox.
+
+    Attributes:
+        ppid: parent process id
+        pid: process id
+        inner: sandbox-specific data
+    """
+
+    ppid: int
+    pid: int
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:279]: hm, now that i think about this: this is the abstract interface for dynamic handles, which means that all implementors *must* be able to provide this data. if we're not sure that its available, then we should rely on implementors to subclass and/or use the `inner` property.

with that in mind, i think we should at least remove ppid. im not sure about using pid/tid. for the time being, lets keep this, with the understanding that as we add support for other sandboxes, we might have to change this interface.
```suggestion
    pid: int
```

File: capa/features/extractors/base_extractor.py [hunk 7]
```diff
@@ -262,3 +262,88 @@ def extract_insn_features(
...
+    Features are grouped mainly into threads that alongside their meta-features are also grouped into 
+    processes (that also have their own features). Other scopes (such as function and file) may also apply 
+    for a specific sandbox.
+
+    This class is not instantiated directly; it is the base class for other implementations.
+    """
+
+    def __init__(self):
+        super().__init__()
+
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:310]:
```suggestion
```

default implementation doesn't have to be specified

**Review by williballenthin — CHANGES_REQUESTED** [review]
two minor suggested changes, then lets merge this!

**Review by williballenthin — APPROVED** [review]

File: capa/features/extractors/base_extractor.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 import abc
 import dataclasses
-from typing import Any, Dict, Tuple, Union, Iterator
+from typing import Any, Dict, Tuple, Union, Iterator, TextIO, BinaryIO
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:11]:
```suggestion
from typing import Any, Dict, Tuple, Union, Iterator
```

File: capa/features/extractors/base_extractor.py [hunk 9]
```diff
@@ -262,3 +262,83 @@ def extract_insn_features(
...
         raise NotImplementedError()
+
+
+@dataclass
+class ProcessHandle:
+    """
+    reference to a process extracted by the sandbox.
+
+    Attributes:
+        ppid: parent process id
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:273]:
```suggestion
```

File: capa/features/extractors/base_extractor.py [hunk 10]
```diff
@@ -262,3 +262,83 @@ def extract_insn_features(
...
+
+    This class is not instantiated directly; it is the base class for other implementations.
+    """
+    @abc.abstractmethod
+    def get_processes(self) -> Iterator[ProcessHandle]:
+        """
+        Yields all the child-processes of a parent one.
+
+        Attributes:
+            ph: parent process
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:312]:
```suggestion
        Enumerate processes in the trace.
```

File: capa/features/extractors/base_extractor.py [hunk 11]
```diff
@@ -262,3 +262,83 @@ def extract_insn_features(
...
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_threads(self, ph: ProcessHandle) -> Iterator[ThreadHandle]:
+        """
+        Yields all the threads that a process created.
+
+        Attributes:
+            ph: parent process
```
Comment (williballenthin) [inline capa/features/extractors/base_extractor.py:332]:
```suggestion
        Enumerate threads in the given process.
```

**Review by williballenthin — CHANGES_REQUESTED** [review]
sorry, more small tweaks. i'll try to merge them directly if i have perms.

File: capa/features/extractors/base_extractor.py (hunk 8 above)
Comment (williballenthin) [reply capa/features/extractors/base_extractor.py:11]: i dont know why our code style linter didn't catch this?

Comment (williballenthin) [reply capa/features/extractors/base_extractor.py]: oh cause it was blocked by the changelog checker

**Review by williballenthin — APPROVED** [review]

**Pull request merged**